R/utils.R

```diff
@@ -259,85 +259,560 @@ rowcol_to_dim <- function(row, col) {
+    }
+    if (identical(col_span, 0L)) {
+      stop("Providing `cols = 0` without an object with dimensions is not supported", "Use `cols = 1`.")
+    }
+
+    if (length(row_span) == 1 && length(col_span) == 1) {
+      # A1
+      row_start <- row_span
+      col_start <- col_span
+      dims <- rowcol_to_dim(row_start, col_start)
+    } else {
+      # A1:B2
+      dims <- rowcol_to_dims(row_span, col_span)
+    }
+    return(dims)
+  }
 
+  # Fix an error I created
+  # if (frow_null)  {
+  #   srow <- srow + 1L
+  # }
+  # if (fcol_null) {
+  #   scol <- scol + 1L
+  # }
+
+
+
+  # After this point, we only cover the case for `x`
+  rows_arg
+  cols_arg
+  if (cnam_null && x_has_named_dims) {
+    if (identical(rows_arg, 0L)) {
+      # message("Use `col_names = TRUE` explicitly to select `x + its column names`",
+      #        "\nBy default, when `x` is specified, `rows` and `cols` are only related to the content of `x`.")
+      # args$col_names <- FALSE
+    } else if (is.null(rows_arg)) {
+      # message(
+      #   "Use `col_names = TRUE` explicitly to select `x + its column names`",
+      #   "\nBy default, when `x` is specified, `rows` and `cols` are only related to the content of `x`."
+      # )
+      args$col_names <- FALSE
+    } else {
+      # message(
+      #   "Use `col_names = TRUE` explicitly to select `x + its column names`",
+      #   "\nBy default, when `x` is specified, `rows` and `cols` are only related to the content of `x`."
+      # )
+      args$col_names <- FALSE
+    }
   }
 
-  rows <- rows + srow
-  cols <- cols + scol
+  col_names <- args$col_names %||% x_has_named_dims
+
+  if (x_present && !col_names && x_has_named_dims && !cnam_null) {
+    if (x_has_colnames) {
+      warning("`x` has column names. Yet, you are asking for `col_names = FALSE`.",
+        "\n ",
+        "\n Consider supplying `x = unname(`input`)`, or use `wb_dims()` without `x` to ensure no errors with `col_names = FALSE`",
+        call. = FALSE
+      )
+    } else {
+      # message("`x` doesn't have col names. assuming there is no name. Supply `col_names = TRUE` only to select rows + column name.")
+    }
+  }
 
-  if (length(rows) == 1 && length(cols) == 1) {
-    # A1
-    dims <- rowcol_to_dim(rows, cols)
+  if (!cnam_null && !x_has_named_dims) {
+    stop("Supplying `col_names` when `x` is a vector is not supported.")
+  }
+  row_names <- args$row_names %||% FALSE
+  assert_class(col_names, "logical")
+  assert_class(row_names, "logical")
+
+  if ((!is.null(rows_arg) && !col_names) && !is.null(cols_arg) && row_names && x_has_named_dims) {
+    warning("The combination of `row_names = TRUE` and `col_names = FALSE` is not recommended.",
+      "unless supplying `cols` and/or `rows`",
+      "`col_names` allows to select the region that contains the data only.",
+      "`row_names` = TRUE adds row numbers if the data doesn't have rownames.",
+      call. = FALSE
+    )
+  }
+  if (!frow_null && identical(srow, -1L)) {
```

Comment:
we could enforce `max(1, srow)`/`max(1, scol)`

Reply:
That would be a good idea. I just want to enforce `rows` `cols` = 0 as a special case with `x`, but it's been a little tricky :)